### ISSUE TITLE
clip_denoised purpose?

### ISSUE THREAD

**Issue by malekinho8** [lucidrains/denoising-diffusion-pytorch#65]
Hello, I was wondering why we have to clamp the output of x_start from predict_start_from_noise( ) when sampling:

```
def p_mean_variance(self, x, t, clip_denoised: bool):

        model_output = self.denoise_fn(x, t)

        if self.objective == 'pred_noise':
            x_start = self.predict_start_from_noise(x, t = t, noise = model_output) # i don't understand this
        elif self.objective == 'pred_x0':
            x_start = model_output
        else:
            raise ValueError(f'unknown objective {self.objective}')

        if clip_denoised:
            x_start.clamp_(-1., 1.)

        model_mean, posterior_variance, posterior_log_variance = self.q_posterior(x_start = x_start, x_t = x, t = t)
        return model_mean, posterior_variance, posterior_log_variance
```

I did not see any mention of imposing this constraint anywhere in the original DDPM paper, and in this other [Google Colab implementation ](https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/annotated_diffusion.ipynb#scrollTo=f70235f8), they do not implement any sort of clamping during the sampling process. Is there a reason why this was done in this code?

**Comment by xenova** [comment]
I believe it is because this implementation converts the output of the model back to images in the sampling function:
https://github.com/lucidrains/denoising-diffusion-pytorch/blob/37334ae82467197ba3df194cae3a85332f5736be/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L535

So, to ensure that the final image is between 0 and 1, you must have that the output from the model (i.e., the denoised image) is between -1 and 1.

**Comment by almson** [comment]
@xenova But that should be a final step, not done on every iteration.

The problem I'm experiencing is that the reverse process doesn't converge toward the mean/variance of the source data. Samples are oversaturated. Removing clipping makes this even worse. I suspect that the clipping is a band-aid for another problem.

**Comment by xenova** [comment]
> @xenova But that should be a final step, not done on every iteration.
> 
> 
> 
> The problem I'm experiencing is that the reverse process doesn't converge toward the mean/variance of the source data. Samples are oversaturated. Removing clipping makes this even worse. I suspect that the clipping is a band-aid for another problem.

Out of curiosity, what beta/noise scheduler are you using? I've had issues in the past with the cosine beta scheduler for some reason.

**Comment by malekinho8** [comment]
> > @xenova But that should be a final step, not done on every iteration.
> > The problem I'm experiencing is that the reverse process doesn't converge toward the mean/variance of the source data. Samples are oversaturated. Removing clipping makes this even worse. I suspect that the clipping is a band-aid for another problem.
> 
> Out of curiosity, what beta/noise scheduler are you using? I've had issues in the past with the cosine beta scheduler for some reason.

I'm not sure about @xenova, but I was using the cosine scheduler and was running into issues with the mean/variance blowing up in the output. In a related [issue](https://github.com/lucidrains/denoising-diffusion-pytorch/issues/58#issuecomment-1186208177), another user found that training models with the pred_x0 objective instead of the default pred_noise objective was actually leading to better samples, and I also found that when clip_denoised was set to False, the end output became oversaturated and exploded away from the -1 to 1 range. Another subtlety to note is that the cosine schedule is not necessarily the best beta schedule to use and it may not actually be the best fit with the simple pred_noise training objective/loss function as used in this code repository:

![Screenshot from 2022-08-01 07-59-35](https://user-images.githubusercontent.com/78977545/182154174-da5a0e3f-49e0-46fb-b22c-a8620b5a8a09.png)

(Taken from Nichol and Dhariwal 2021)

**Comment by yanglibo0512** [comment]
> @malekinho8 Hi, do we have the answer of the problem? I am confused of clip_denoised purpose and the range of it. I noticed that in improved DDPM, they scaled the image intensity value to -1 and 1. I'm not sure if this is relevant to the [-1,1] range set in clip_dennoised function.
> 
> Looking forward to your reply.

I also found this problem, so is it related to the range of data? If the data range is 0-1, does the clamp range here also need to be changed accordingly?

**Comment by lucidrains** [comment]
@yanglibo0512 the data is normalized to -1 to 1 [here](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L585)

**Comment by yanglibo0512** [comment]
> @yanglibo0512 the data is normalized to -1 to 1 [here](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L585)

I see, I want to know whether the range of clamp here corresponds to the range of data?

**Comment by AwayGithub** [comment]
@malekinho8 @xenova @almson @GYDDHPY @yanglibo0512 
Update: I've uncovered some new insights. https://github.com/lucidrains/denoising-diffusion-pytorch/issues/47#issuecomment-2768297917 are some of my observations that may help clarify this issue.

In summary:

- For DDPM, clamping is necessary at high sampling steps (around 1000) to ensure stability.
- As the step count decreases, clamping can be removed—though the exact reason remains unclear—but performance gradually degrades with fewer steps.
- Since DDIM operates at significantly fewer steps than DDPM, clamping can safely be omitted.
